test/validation/api/ipsec/ipsec_test_in.c

```diff
@@ -1727,6 +1768,254 @@ static void test_ipsec_sa_print(void)
 	ipsec_sa_destroy(in_sa);
 }
 
+static void test_in_ipv4_esp_reass_success_two_frags(odp_ipsec_sa_t out_sa,
+						     odp_ipsec_sa_t in_sa)
+{
+	ipsec_test_part test_out[MAX_FRAGS], test_in[MAX_FRAGS];
+
+	part_prep_esp(test_in, 2, false);
+
+	test_in[0].pkt_in = &pkt_ipv4_udp_p4_f1;
+	test_in[1].pkt_in = &pkt_ipv4_udp_p4_f2;
+
+	ipsec_out_verify_send(test_in, out_sa, 2);
+
+	part_prep_plain(test_out, 1, 1, false, true);
+	test_out[0].out[0].pkt_res = &pkt_ipv4_udp_p4;
+
+	ipsec_in_recv_verify(test_out, in_sa, 1);
+}
+
+static void test_in_ipv4_esp_reass_success_four_frags(odp_ipsec_sa_t out_sa,
+						      odp_ipsec_sa_t in_sa)
+{
+	ipsec_test_part test_out[MAX_FRAGS], test_in[MAX_FRAGS];
+
+	part_prep_esp(test_in, 4, false);
+
+	test_in[0].pkt_in = &pkt_ipv4_udp_p5_f1;
+	test_in[1].pkt_in = &pkt_ipv4_udp_p5_f2;
+	test_in[2].pkt_in = &pkt_ipv4_udp_p5_f3;
+	test_in[3].pkt_in = &pkt_ipv4_udp_p5_f4;
+
+	ipsec_out_verify_send(test_in, out_sa, 4);
+
+	part_prep_plain(test_out, 1, 1, false, true);
+	test_out[0].out[0].pkt_res = &pkt_ipv4_udp_p5;
+
+	ipsec_in_recv_verify(test_out, in_sa, 1);
+}
+
+static void test_in_ipv4_esp_reass_success_two_frags_ooo(odp_ipsec_sa_t out_sa,
+							 odp_ipsec_sa_t in_sa)
+{
+	ipsec_test_part test_out[MAX_FRAGS], test_in[MAX_FRAGS];
+
+	part_prep_esp(test_in, 2, false);
+
+	test_in[1].pkt_in = &pkt_ipv4_udp_p4_f2;
+	test_in[0].pkt_in = &pkt_ipv4_udp_p4_f1;
+
+	ipsec_out_verify_send(test_in, out_sa, 2);
+
+	part_prep_plain(test_out, 1, 1, false, true);
+	test_out[0].out[0].pkt_res = &pkt_ipv4_udp_p4;
+
+	ipsec_in_recv_verify(test_out, in_sa, 1);
+}
+
+static void test_in_ipv4_esp_reass_success_four_frags_ooo(odp_ipsec_sa_t out_sa,
+							  odp_ipsec_sa_t in_sa)
+{
+	ipsec_test_part test_out[MAX_FRAGS], test_in[MAX_FRAGS];
+
+	part_prep_esp(test_in, 4, false);
+
+	test_in[0].pkt_in = &pkt_ipv4_udp_p5_f2;
+	test_in[1].pkt_in = &pkt_ipv4_udp_p5_f4;
+	test_in[2].pkt_in = &pkt_ipv4_udp_p5_f3;
+	test_in[3].pkt_in = &pkt_ipv4_udp_p5_f1;
+
+	ipsec_out_verify_send(test_in, out_sa, 4);
+
+	part_prep_plain(test_out, 1, 1, false, true);
+	test_out[0].out[0].pkt_res = &pkt_ipv4_udp_p5;
+
+	ipsec_in_recv_verify(test_out, in_sa, 1);
+}
+
+static void test_in_ipv4_esp_reass_incomp_overlap(odp_ipsec_sa_t out_sa,
+						  odp_ipsec_sa_t in_sa)
+{
+	ipsec_test_part test_out[MAX_FRAGS], test_in[MAX_FRAGS];
+
+	part_prep_esp(test_in, 3, false);
+
+	test_in[0].pkt_in = &pkt_ipv4_udp_p4_f1;
+	test_in[1].pkt_in = &pkt_ipv4_udp_p4_f1;
+	test_in[2].pkt_in = &pkt_ipv4_udp_p4_f2;
+
+	ipsec_out_verify_send(test_in, out_sa, 3);
+
+	part_prep_plain(test_out, 1, 3, false, false);
+
+	test_out[0].out[0].pkt_res = &pkt_ipv4_udp_p4_f1;
+	test_out[0].out[1].pkt_res = &pkt_ipv4_udp_p4_f1;
+	test_out[0].out[2].pkt_res = &pkt_ipv4_udp_p4_f2;
+
+	ipsec_in_recv_verify(test_out, in_sa, 1);
```

Comment:
This makes the incorrect assumption that incomplete reassembly results contain the original fragments. The API allows an implementation to perform partial reassembly and e.g. return just single fragment. Also, my take on the API is that an implementation is allowed to drop overlapping fragments without ever delivering them to an application.

And similar comment to the other functions testing incomplete reassembly.

Reply:
I'll drop these incomplete reassembly tests.

Reply:
Or maybe just have them check that the incomplete reassembly status gets reported, without checking the returned fragments?

Reply:
Removed the tests for now. Will propose further incomplete tests as part of separate PR.